### PR TITLE
extmod/vfs_fat: Object-oriented encapsulation of FatFs VFS.

### PR DESCRIPTION
**Pull request by pfalcon**
This implements OO interface based on existing fsusermount code and with
minimal changes to it, to serve as a proof of concept of OO interface.

Examle of usage:

bdev = RAMFS(48)
uos.VfsFat.mkfs(bdev)
vfs = uos.VfsFat(bdev, "/ramdisk")
f = vfs.open("foo", "w")
f.write("hello!")
f.close()

### CODE REVIEW

**Comment by pfalcon** [comment]
Simple as it seems (well, majority of changes required for this were merged in my previous commits). Changes to enable this for unix port will follow.

**Comment by pfalcon** [comment]
Pushed commit to enable on unix.

File: extmod/vfs_fat.c [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    { MP_ROM_QSTR(MP_QSTR_mkfs), MP_ROM_PTR(&fat_vfs_mkfs_obj) },
+    { MP_ROM_QSTR(MP_QSTR_open), MP_ROM_PTR(&fat_vfs_open_obj) },
+};
+STATIC MP_DEFINE_CONST_DICT(fat_vfs_locals_dict, fat_vfs_locals_dict_table);
+
+const mp_obj_type_t mp_fat_vfs_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_VfsFat,
+    .make_new = fat_vfs_make_new,
+    .locals_dict = (void*)&fat_vfs_locals_dict,
```
Comment (dpgeorge) [inline extmod/vfs_fat.c]: void* can be mp_obj_dict_t*

File: unix/modos.c [hunk 2]
```diff
@@ -228,6 +236,14 @@ STATIC const mp_rom_map_elem_t mp_module_os_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_getenv), MP_ROM_PTR(&mod_os_getenv_obj) },
     { MP_ROM_QSTR(MP_QSTR_mkdir), MP_ROM_PTR(&mod_os_mkdir_obj) },
     { MP_ROM_QSTR(MP_QSTR_ilistdir), MP_ROM_PTR(&mod_os_ilistdir_obj) },
+    #if MICROPY_FSUSERMOUNT
+    { MP_ROM_QSTR(MP_QSTR_vfsmount), MP_ROM_PTR(&fsuser_mount_obj) },
+    { MP_ROM_QSTR(MP_QSTR_vfsumount), MP_ROM_PTR(&fsuser_umount_obj) },
+    { MP_ROM_QSTR(MP_QSTR_vfsmkfs), MP_ROM_PTR(&fsuser_mkfs_obj) },
```
Comment (dpgeorge) [inline unix/modos.c]: These names are quite a mouthful... maybe underscore them, ie vfs_mount?

File: extmod/vfs_fat.c (hunk 1 above)
Comment (pfalcon) [inline extmod/vfs_fat.c]: fixed

File: unix/modos.c (hunk 2 above)
Comment (pfalcon) [inline unix/modos.c]: Well, this tries to follow existing "almost core" function naming like listdir().

**Comment by dpgeorge** [comment]
I'm confused by how mounting happens in this picture.  In the non-OO picture I would expect to be able to do the following:
```python
bdev = RAMBlockDevice(64)
os.mkfs(bdev, type='fatfs')
os.mount(bdev, '/ramfs')
f = open('/ramfs/test')
...
```
With the OO style I was thinking it would be something like:
```python
bdev = RAMBlockDevice(64)
fatfs = FatFS(bdev)
fatfs.mkfs()
f = fatfs.open('test')
...
```
In this OO picture, if I then create a second FatFS object it would be completely independent to the first.  So there should be no mention of mount point in this picture.  For the concept of mounting, you need a higher-level object which is a container filesystem (let's call it vfs):
```python
vfs = VFS()
vfs.mount(fatfs, '/ramfs')
vfs.mount(fatfs2, '/ramfs2')
f = vfs.open('/ramfs/test')
```
The way you implemented it, you need to specify the mount point when creating a VfsFat object.  So what is VfsFat: is it a single filesystem backed by a single block device, or is it a set of filesystems?

**Comment by dpgeorge** [comment]
In the OO picture, a function/method like "open" or "unlink" is implemented per filesystem type (eg fatfs, ext2, etc).  But I think that is unnecessary duplication of code because all filesystem implementations will need the same "glue" code for open and unlink, etc, and just have different backends to do the actual operation.  I was thinking that it would be similar to the stream API: there is a single function (eg open, unlink) for the entry point and it does the boilerplate stuff (eg checking args, raising exceptions), and then this function calls through to the underlying implementation.

File: unix/modos.c (hunk 2 above)
Comment (dpgeorge) [inline unix/modos.c]: But vfs is already an acronym (unlike "list" and "dir") so is harder to read (and it's not familiar like mkdir is).  But I hope that these ugly names are only for unix and baremetal ports can chose to use mount/umount/mkfs :)

**Comment by pfalcon** [comment]
> So there should be no mention of mount point in this picture.

Heh, in the commit message I wrote, line `vfs = uos.VfsFat(bdev, "/ramdisk")` was preceded by Python-style comment: "In the ideal OO implementation, there wouldn't be 2nd arg. But as OO interface is just shim on top ox fsusermount, it is there." Well, as Python-style comments are also git commit style comment, it ate it!

**Comment by pfalcon** [comment]
> So what is VfsFat: is it a single filesystem backed by a single block device, or is it a set of filesystems?

It is single filesystem backed by a single block device, with the weight of technical debt and mis-designed FatFs implementation ;-). Beyond stray args here and there, it has interface as expected!

**Comment by pfalcon** [comment]
> In the OO picture, a function/method like "open" or "unlink" is implemented per filesystem type (eg fatfs, ext2, etc).

In OO picture, it is virtual method, and per-FS implementation of that method is "underlying implementation".

File: unix/modos.c (hunk 2 above)
Comment (pfalcon) [inline unix/modos.c]: Actually, my several-months-ago idea was to propose baremetals to adopt them too ;-). But yep, nowadays I'm looking towards dropping them completely in favor of OO instantiation.

**Comment by dpgeorge** [comment]
Did you try to make 2 (or more) RAMFS devices, and then VfsFat instances, at the same time and verify that they work as expected?  Fatfs may access only the first instance (due to CurVol).

**Comment by pfalcon** [comment]
> Did you try to make 2 (or more) RAMFS devices, and then VfsFat instances, at the same time and verify that they work as expected? Fatfs may access only the first instance (due to CurVol).

Just to clarify: the aim of these changes was to apply minimal changes to fsusermount, and zero changes to unix port code, while making VFS functionality available to unix port, to establish a testing baseline for it, and make porting to other targets easy. When I started, there was no support for multiple user volumes, and after you added it, I just ported my existing changes over it and submitted them.

You're right, with the current (mis)design of FatFs, 2 or more volumes would work only if user passes absolute path with the same prefix as was used when instantiating VfsFat object. Again, this is a known issue (design defect) of FatFs. A properly designed VFS library wouldn't have that problem, and my framework readily allows to integrate it in uPy. We can also fix that for FatFs (by patching it, as working that around externally is quite boring and resource-wasting). But that's beyond the immediate aim of establishing proven (==testable) reusability baseline. We should not treat each feature added as having final interface/implementation, or it will make adding many features, as e.g. required for #1806, quite problematic. We should consider this "PoC" feature, to be finalized, say, to 1.8.

Can you in the meantime confirm that PD_FLASH, etc. are no longer relevant and should be removed from lib/fatfs/ ? I then set number of FatFs volumes for unix to 1.

**Comment by pfalcon** [comment]
Nanbox opened Pandora box in stmhal...

**Comment by pfalcon** [comment]
Ok, I changed names to use "vfs_" as was suggested. And with a testcase added (which required implementing "in" for bytearrays!), finally going to merge this - this already proved to take much more effort than many other things, and that's only a beginning ;-).

**Comment by dpgeorge** [comment]
> We can also fix that for FatFs (by patching it, as working that around externally is quite boring and resource-wasting).

Yes, I want to do this.  I'm thinking of making a clone of fatfs library in a separate repository and applying the changes to make it more OO friendly (ie each function takes a bdev argument upon which the fatfs actually operates; this would simplify a lot of the chdir/getvol code in fatfs).  Actually, I'd like to make such a repo more general and include other filesystem types, all with the same C API for easy plugging.

> Can you in the meantime confirm that PD_FLASH, etc. are no longer relevant and should be removed from lib/fatfs/ ?

Yes, those defines should go away now.  Except that cc3200 seems to use them... I guess it needs some attention.

> I then set number of FatFs volumes for unix to 1.

Where does unix port get the VolToPart table from?  Remember that that table must be larger than the maximum number of volumes.

File: tests/extmod/vfs_fat_ramdisk.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    def writeblocks(self, n, buf):
+        #print("writeblocks(%s, %x)" % (n, id(buf)))
+        for i in range(len(buf)):
+            self.data[n*512+i] = buf[i]
+
+    def sync(self):
+        pass
+
+    def count(self):
```
Comment (dpgeorge) [inline tests/extmod/vfs_fat_ramdisk.py]: This is using old-style protocol.  Might be good to change it to new ioctl style so other people who copy the code also use the new style.

Comment (pfalcon) [inline tests/extmod/vfs_fat_ramdisk.py]: Already done.

**Comment by pfalcon** [comment]
> to make it more OO friendly (ie each function takes a bdev argument upon which the fatfs actually operates

Good idea. Except to be exact, it should be not the underlying blockdev, but FatFs' own structure (called FATFS).

> Where does unix port get the VolToPart table from? Remember that that table must be larger than the maximum number of volumes.

unix doesn't enable support for multiple partitions per volume.

**Pull request merged**